File: java/spring-resilience4j/spring-resilience4j-api-a/src/main/java/com/bmarques/springresilience4japia/revenue/RevenueService.java
package com.bmarques.springresilience4japia.revenue;

import com.bmarques.springresilience4japia.invoice.Invoice;
import com.bmarques.springresilience4japia.invoice.InvoiceClient;
import io.github.resilience4j.circuitbreaker.annotation.CircuitBreaker;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
@Slf4j
public class RevenueService {

    @Autowired
    private InvoiceClient invoiceClient;

    @CircuitBreaker(name = "invoiceApi", fallbackMethod = "fallToGetTotalRevenue")
    public Double getTotalRevenue() {
        List<Invoice> invoices = invoiceClient.getInvoices();

        return invoices.stream()
                .map(Invoice::getValue)
                .mapToDouble(Double::doubleValue)
                .sum();
    }

    public Double fallToGetTotalRevenue(Throwable e) {
        log.info("Fallback activated");
        return 0D;
    }
}

File: kotlin/spring-resilience4j/spring-resilience4j-api-a/src/main/kotlin/com/bmarques/springresilience4japia/revenue/RevenueService.kt
package com.bmarques.springresilience4japia.revenue

import com.bmarques.springresilience4japia.invoice.Invoice
import com.bmarques.springresilience4japia.invoice.InvoiceClient
import io.github.resilience4j.circuitbreaker.annotation.CircuitBreaker
import org.springframework.stereotype.Service

@Service
class RevenueService(private val invoiceClient: InvoiceClient) {

    @CircuitBreaker(name = "invoiceApi", fallbackMethod = "fallToGetTotalRevenue")
    fun getTotalRevenue(): Double {
        val invoices = invoiceClient.getInvoices()
        return invoices.stream()
                .map(Invoice::value)
                .mapToDouble { obj: Double -> obj }
                .sum()
    }

    fun fallToGetTotalRevenue(e: Throwable): Double = 0.0
}